mrmustard/lab_dev/states/dm.py

```diff
@@ -359,6 +360,249 @@ def quadrature_distribution(self, quad: RealVector, phi: float = 0.0) -> Complex
         quad = math.tile(quad, (1, 2))
         return self.quadrature(quad, phi)
 
+    def stellar_decomposition(self, core_modes: Collection[int]):
+        r"""
+        Computes the formal stellar decomposition for the DM.
+
+        Args:
+            core_modes: The set of modes defining core variables.
+
+        Returns:
+            core: The core state (`DM`)
+            phi: The Gaussian `Map` performing the stellar decomposition (not necessarily CPTP).
+
+        Raises:
+            ValueError: if the state is non-Gaussian.
+
+        Note:
+            This method pulls out the map ``phi`` from the given state on the given modes, so that
+            the remaining state is a core state. Formally, we have
+            .. math::
+
+                \rho = (\phi\otimes\mathcal I) \rho_{\mathrm{core}}
+
+            where the map :math:`phi` acts on the given `core_modes` only.
+            Core states have favorable properties in the Fock representation
+            e.g., being sparse.
+
+        .. code-block::
+            >>> from mrmustard.lab_dev import DM
+
+            >>> rho = DM.random([0,1])
+            >>> core, phi = rho.stellar_decomposition([0])
+            >>> assert rho == core >> phi
+        """
+        A, b, c = self.ansatz.triple
+        A = A[-1]
+        b = b[-1]
+        if c.shape != (1,):
+            raise ValueError(
+                f"The stellar decomposition only applies to Gaussian states. The given state has a polynomial of size {c.shape}."
+            )
+
+        m_modes = A.shape[-1] // 2
+
+        mode_to_idx = {q.mode: q.index for q in self.wires.quantum_wires}
+        core_bra_indices = [
+            (i - m_modes) if i >= m_modes else i for i in (mode_to_idx[j] for j in core_modes)
+        ]
+        core_ket_indices = [(i + m_modes) for i in core_bra_indices]
+        core_indices = core_bra_indices + core_ket_indices
+        remaining_indices = [i for i in range(2 * m_modes) if i not in core_indices]
+        new_order = math.astensor(core_indices + remaining_indices)
```

Comment:
Perhaps it's easier to reorder the ansatz directly using the reorder() method. You could do this:
```python
other_modes = [m for m in self.modes if m not in core_modes]
core_indices = self.wires[core_modes].indices
other_indices = self.wires[other_modes].indices
new_order = core_indices + other_indices
A,b,c = self.ansatz.reorder(new_order)
```
Unless I misunderstood, this should be what you want.